File: Questions/3. Longest Substring Without Repeating Characters/main.js
function lengthOfLongestSubstring(s) {
    const map = new Map()
    let left = 0, right = 0
    const n = s.length
    let ret = 0

    while (right < n) {
        const c = s[right]
        map.set(c, (map.get(c) || 0) + 1)
        
        while (left < right && map.get(c) > 1) {
            const lc = s[left]
            map.set(lc, map.get(lc) - 1)
            left++
        }
        ret = Math.max(ret, right - left + 1)
        right++
    }
    return ret
}

console.log(lengthOfLongestSubstring("abcabcbb"));

File: Questions/3. Longest Substring Without Repeating Characters/main.py
from collections import defaultdict
def length_of_longest_substring(s: str) -> str:
    count = defaultdict(int)
    left = right = 0
    ret = 0

    while right < len(s):
        count[s[right]] += 1
        while left < right and count[s[right]] > 1:
            count[s[left]] -= 1
            left += 1
        
        ret = max(ret, right - left + 1)
        right += 1

    return ret

assert length_of_longest_substring("abcabcbb") == 3
print(length_of_longest_substring("abcabcbb"))